addons/controllers/cpi/vspherecpiconfig_controller.go

```diff
@@ -80,34 +80,45 @@ var vsphereCPIProviderServiceAccountAggregatedClusterRole = &rbacv1.ClusterRole{
 
 // Reconcile the VSphereCPIConfig CRD
 func (r *VSphereCPIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := r.Log.WithValues("VSphereCPIConfig", req.NamespacedName)
+	logger := r.Log.WithValues("VSphereCPIConfig", req.NamespacedName)
 
-	log.Info("Start reconciliation for VSphereCPIConfig")
+	logger.Info("Start reconciliation for VSphereCPIConfig")
 
 	// fetch VSphereCPIConfig resource
 	cpiConfig := &cpiv1alpha1.VSphereCPIConfig{}
 	if err := r.Client.Get(ctx, req.NamespacedName, cpiConfig); err != nil {
 		if apierrors.IsNotFound(err) {
-			log.Info("VSphereCPIConfig resource not found")
+			logger.Info("VSphereCPIConfig resource not found")
 			return ctrl.Result{}, nil
 		}
-		log.Error(err, "Unable to fetch VSphereCPIConfig resource")
+		logger.Error(err, "Unable to fetch VSphereCPIConfig resource")
 		return ctrl.Result{}, err
 	}
 
 	// deep copy VSphereCPIConfig to avoid issues if in the future other controllers where interacting with the same copy
 	cpiConfig = cpiConfig.DeepCopy()
 
-	cluster, err := r.getOwnerCluster(ctx, cpiConfig)
+	cluster, err := cutil.GetOwnerCluster(ctx, r.Client, cpiConfig, req.Namespace, constants.CPIDefaultRefName)
+	if err != nil {
+		if apierrors.IsNotFound(err) && cluster != nil {
+			logger.Info(fmt.Sprintf("'%s/%s' is listed as owner reference but could not be found",
+				cluster.Namespace, cluster.Name))
+			return ctrl.Result{}, err
+		}
+		logger.Error(err, "could not determine owner cluster")
+		return ctrl.Result{}, err
+	}
+
 	if cluster == nil {
-		return ctrl.Result{}, err // no need to requeue if cluster is not found
+		return ctrl.Result{}, err // retry until corresponding cluster is found
```

Comment:
Will error be there if cluster is nil ?

Reply:
yes. cutil.GetOwner aways returns an error with cluster=nil, but even without knowing the internals of cutil.GetOwnerCluster,  if cluster is nil && and err = nil (which is hard to think of a case that would happe), it would just flag the request as done which is what we want if err == nil .

Reply:
actully after looking at GetOwner description, and code, it will never return cluster=nil & err = nil 
the only time cluster =nil is when err != nil, so we can skip this altogehter.